From 74cf148e8ec39512fb026f944005e1f474399065 Mon Sep 17 00:00:00 2001
From: Scott Mayhew <smayhew@redhat.com>
Date: Fri, 13 Jun 2014 15:26:51 -0400
Subject: [fs] nfs: Apply NFS_MOUNT_CMP_FLAGMASK to nfs_compare_remount_data()

Message-id: <1402673211-45912-1-git-send-email-smayhew@redhat.com>
Patchwork-id: 83909
O-Subject: [RHEL 7.1 PATCH] nfs: Apply NFS_MOUNT_CMP_FLAGMASK to nfs_compare_remount_data()
Bugzilla: 1103805
RH-Acked-by: Steve Dickson <SteveD@redhat.com>
RH-Acked-by: Sachin Prabhu <sprabhu@redhat.com>
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

Fixes bug 1103805.

Those flags are obsolete and checking them can incorrectly cause
remount operations to fail.

Signed-off-by: Scott Mayhew <smayhew@redhat.com>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
(cherry picked from commit c8e470280a5c875468dd4eb41d8c4b44c87ebbcf)
Signed-off-by: Scott Mayhew <smayhew@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/super.c b/fs/nfs/super.c
index d5fb771..6766eba 100644
--- a/fs/nfs/super.c
+++ b/fs/nfs/super.c
@@ -2179,11 +2179,23 @@ out_no_address:
  return -EINVAL;
 }
 
+#define NFS_MOUNT_CMP_FLAGMASK ~(NFS_MOUNT_INTR \
+  | NFS_MOUNT_SECURE \
+  | NFS_MOUNT_TCP \
+  | NFS_MOUNT_VER3 \
+  | NFS_MOUNT_KERBEROS \
+  | NFS_MOUNT_NONLM \
+  | NFS_MOUNT_BROKEN_SUID \
+  | NFS_MOUNT_STRICTLOCK \
+  | NFS_MOUNT_UNSHARED \
+  | NFS_MOUNT_NORESVPORT \
+  | NFS_MOUNT_LEGACY_INTERFACE)
+
 static int
 nfs_compare_remount_data(struct nfs_server *nfss,
     struct nfs_parsed_mount_data *data)
 {
- if (data->flags != nfss->flags ||
+ if ((data->flags ^ nfss->flags) & NFS_MOUNT_CMP_FLAGMASK ||
      data->rsize != nfss->rsize ||
      data->wsize != nfss->wsize ||
      data->version != nfss->nfs_client->rpc_ops->version ||
@@ -2344,18 +2356,6 @@ void nfs_clone_super(struct super_block *sb, struct nfs_mount_info *mount_info)
   nfs_initialise_sb(sb);
 }
 
-#define NFS_MOUNT_CMP_FLAGMASK ~(NFS_MOUNT_INTR \
-  | NFS_MOUNT_SECURE \
-  | NFS_MOUNT_TCP \
-  | NFS_MOUNT_VER3 \
-  | NFS_MOUNT_KERBEROS \
-  | NFS_MOUNT_NONLM \
-  | NFS_MOUNT_BROKEN_SUID \
-  | NFS_MOUNT_STRICTLOCK \
-  | NFS_MOUNT_UNSHARED \
-  | NFS_MOUNT_NORESVPORT \
-  | NFS_MOUNT_LEGACY_INTERFACE)
-
 static int nfs_compare_mount_options(const struct super_block *s, const struct nfs_server *b, int flags)
 {
  const struct nfs_server *a = s->s_fs_info;
-- 
1.7.1